OpenTelemetry (S3M) App Development apps (#1798)
SumoLogic/sumologic-documentation: docs/integrations/app-development/opentelemetry/puppet-opentelemetry.md

 logsjson['transaction_uuid'] = obj['transaction_uuid'];
 logsjson['environment'] = obj['environment'];
 puts logsjson.to_json
 };"
echo "$f conversion status:" $? >> "$log_file_name"
date "+%d-%b-%Y %H:%M:%S" > last_run_date.tmp
fi
done
echo "Conversion Completed. End Time: $(date)" >> "$log_file_name"
```

Copy this script to `/opt/puppetlabs/server/data/puppetserver/reports` with the name `puppetReport.sh`. If Puppet is installed in a different location, copy the script to the correct path.

Configure a cron job to trigger the python script using crontab. Frequency of this job can be set following the instructions from [here](https://www.python-engineer.com/posts/cron-jobs-for-python/#crontab). Here is the command which needs to be configured as part of cron to trigger the script:
```sh
<frequency_expression> /opt/puppetlabs/server/data/puppetserver/reports/puppetReport.sh
```
Please modify the location of the `puppetReport.sh` if required in the above command. The execution of the script above will generate a log file named `puppet_rpt_conversion.log`. The path to this log file needs to be provided during the app installation.

## Collection configuration and App installation

> As part of the setting up the collection process and app installation user can select the App from App Catalog and click on **Install App**. Please follow the steps below.

Author: ```suggestion
{@import ../../../reuse/opentelemetry/config-app-install.md}
```

---
id: puppet-opentelemetry
title: Puppet - OpenTelemetry Collector
sidebar_label: Puppet - OTel Collector
description: Learn about the Sumo Logic OpenTelemetry App for Puppet.
---

import useBaseUrl from '@docusaurus/useBaseUrl';
import Tabs from '@theme/Tabs';
import TabItem from '@theme/TabItem';

> <img src={useBaseUrl('img/integrations/app-development/puppet.png')} alt="Thumbnail icon" width="50"/> <img src={useBaseUrl('img/send-data/otel-color.svg')} alt="Thumbnail icon" width="45"/>

Reviewer: This image link is broken when I run it in a local build. Verify that it works in your system.

Puppet is a software configuration management tool. Puppet can provision infrastructure and enforce desired configurations across new and existing servers. The Sumo Logic App for Puppet helps you monitor Puppet events. Puppet logs are sent to Sumo Logic through OpenTelemetry [filelog receiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/filelogreceiver).

<img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/Puppet-OpenTelemetry/Puppet-Schematics.png')} alt="Schematics" />

## Fields creation in Sumo Logic for Puppet

Following are the tags which will be created as part of Puppet App install if not already present. 

- **`sumo.datasource`** - Has a fixed value of **Puppet**.

> ## Prerequisite

Reviewer: ```suggestion
## Prerequisites
```

This section provides instructions for configuring log collection for Puppet running on a non-Kubernetes environment for the Sumo Logic App for Puppet.

Follow the instructions to set up log collection:

We use OpenTelemetry collector to gather the following data from Puppet:

> -   Puppet Server logs. For more information about the logs location, see [Puppet Server Logs](https://puppet.com/docs/puppetserver/5.3/config_file_logbackxml.html).

Reviewer 2: ```suggestion
-   **Puppet Server logs** :  For more information about the logs location, see [Puppet Server Logs](https://puppet.com/docs/puppetserver/5.3/config_file_logbackxml.html).
```

> -   Puppet Server Access logs. For more information about the logs location, see [Puppet Server Logs](https://puppet.com/docs/puppetserver/5.3/config_file_logbackxml.html).

Reviewer 2: ```suggestion
-   **Puppet Server Access logs**. For more information about the logs location, see [Puppet Server Logs](https://puppet.com/docs/puppetserver/5.3/config_file_logbackxml.html).
```

> -   Puppet Reports. Puppet generates reports in YAML format. SumoLogic supports report format 10. This is the format of reports output by Puppet versions 5.5.3 and newer. It is backward compatible with report format 9 (in Puppet versions 5.5.0 to 5.5.2). For more information about the puppet reports, see [Puppet Reports](https://puppet.com/docs/puppet/5.5/format_report.html).

Reviewer 2: Sure. No problem

The default Puppet Server Access log file is: `/var/log/puppetlabs/puppetserver/puppetserver-access.log`. If your Puppet Server Access logs are located elsewhere, please note the path this will be used later while app installation.

> Puppet reports are in YAML format. They must be converted into JSON format before Sumo ingests them. You can convert the YAML files using the sumo provided shell script below:

Reviewer: ```suggestion
Puppet reports are in YAML format. They must be converted into JSON format before Sumo ingests them. You can convert the YAML files using the Sumo-provided shell script below:
```

 logsjson['host'] = obj['host'];
 logsjson['transaction_uuid'] = obj['transaction_uuid'];
 logsjson['environment'] = obj['environment'];
 puts logsjson.to_json
 };"
echo "$f conversion status:" $? >> "$log_file_name"
date "+%d-%b-%Y %H:%M:%S" > last_run_date.tmp
fi
done
echo "Conversion Completed. End Time: $(date)" >> "$log_file_name"
```

Please copy this script to `/opt/puppetlabs/server/data/puppetserver/reports` with the name puppetReport.sh. If Puppet is installed in a different location, copy the script to the correct path.

Configure a cron job to trigger the python script using crontab. Frequency of this job can be set following the instructions from [here](https://www.python-engineer.com/posts/cron-jobs-for-python/#crontab). Here is the command which needs to be configured as part of cron to trigger the script:
```sh
<frequency_expression> /opt/puppetlabs/server/data/puppetserver/reports/puppetReport.sh
```
Please modify the location of the `puppetReport.sh` if required in the above command. The execution of the script above will generate a log file named `puppet_rpt_conversion.log`. The path to this log file needs to be provided during the app installation.

> ## Collection configuration & App installation

Reviewer: ```suggestion
## Collection configuration and App installation
```

As part of the setting up the collection process and app installation user can select the App from App Catalog and click on **Install App**. Please follow the steps below.

### Step 1: Set up Collector

> If you want to use an existing Otel Collector then this step can be skipped by selecting the option of using an existing Collector.

Reviewer 2: Please follow the same convention as in other docs. You can refer any db or webserver based logs.

If you want to create a new Collector please select **Add a new Collector** option.

Select the platform for which you want to install the Sumo OpenTelemetry Collector.

This will generate a command which can be executed in the machine which needs to be monitored. Once executed it will install the Sumo Logic OpenTelemetry Collector agent.

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/Puppet-OpenTelemetry/Puppet-Collector.png')} alt="Collector" />

Reviewer: This image link is broken when I run it in a local build. Verify that it works in your system.

### Step 2: Configure integration

In this step we will be configuring the yaml required for Puppet collection.

> Path of the log file configured to capture puppet access logs and puppet report are needed to be given here - please refer to the prerequisite section of this document.

Reviewer: ```suggestion
Path of the log file configured to capture puppet access logs and puppet report are needed to be given here. Refer to the Prerequisites section of this document.
```

Click on the **Download YAML File** button to get the yaml file.

> <img src={useBaseUrl('https://sumologic-app-data-v2.s3.amazonaws.com/dashboards/Puppet-OpenTelemetry/Puppet-YAML.png')} alt="YAML" />

Reviewer: This image link is broken when I run it in a local build. Verify that it works in your system.

### Step 3: Sending logs to Sumo

Once you have the yaml file downloaded in step 2, follow the below steps based on the platform of the machine:

<Tabs
  className="unique-tabs"
  defaultValue="Linux"
  values={[
    {label: 'Linux', value: 'Linux'},
    {label: 'Windows', value: 'Windows'},
    {label: 'macOS', value: 'macOS'},
  ]}>

<TabItem value="Linux">

1.  Copy the yaml file to `/etc/otelcol-sumo/conf.d/` folder in the Puppet instance which needs to be monitored.

> 2.  Place Env file in the following directory

Reviewer: ```suggestion
2.  Place Env file in the following directory:
```

```sh
/etc/otelcol-sumo/env/
```

> 3.  restart the collector using

Reviewer: ```suggestion
3.  Restart the collector using:
```

```sh
 sudo systemctl restart otelcol-sumo
```

</TabItem>
<TabItem value="Windows">

1.  Copy the yaml file to **`C:\ProgramData\Sumo Logic\OpenTelemetry Collector\config\conf.d`** folder in the machine which needs to be monitored.
2.  Restart the collector using 
```sh
Restart-Service -Name OtelcolSumo
```

</TabItem>

> <TabItem value="macOS">

Reviewer: ```suggestion
<TabItem value="Mac">
```

1.  Copy the yaml file to /etc/otelcol-sumo/conf.d/ folder in the Puppet instance which needs to be monitored.

> 2.  Restart the otelcol-sumo process using the below command 

Reviewer: ```suggestion
2.  Restart the otelcol-sumo process using the below command:
```
